skimage/morphology/rolling_ball.py

```diff
@@ -0,0 +1,178 @@
+import numpy as np
+
+from skimage.util import invert
+
+
+def rolling_ball(image, radius=50, white_background=False):
+    """Perform background subtraction using the rolling ball method.
+
+    The rolling ball filter is a segmentation method that aims to separate the
+    background from a grayscale image in case of uneven exposure. It is
+    frequently used in biomedical image processing and was first proposed by
+    Stanley R. Sternberg (1983) in the paper Biomedical Image Processing [1]_.
+
+    Parameters
+    ----------
+    image : ndarray
+        The image to be filtered.
+    radius : float or tuple, optional
+        The radius of the ball that is used as the structuring element. If
+        ``radius`` is a tuple it must be of the form
+        ``(space_radius, intensity_radius)``, where `space_radius` indicates
```

Comment:
Why not directly use the `space_radius` and `intensity_radius` instead of using this conditional definition of `radius`?

Reply:
I'd say it's a reasonable API if most of the time `space_radius` is expected to be the same as `intensity_radius`.

Reply:
In this case we just have to default them to the same value :)

Reply:
Fine with me.

Reply:
The reason why I have a single `radius` parameter is user expectation. If somebody (end-user) that doesn't know much about CV - say your friendly neighborhood biologist coming from ImageJ - wants to use this function, they would expect to specify a single `radius`. It's what the paper does, and also what all implementations I've seen for this do.

To make the algorithm work with arbitrary dtypes, however, it needs to be generalized to using spheroids instead of balls/spheres; hence, the introduction of the two parameters `(space_radius, intensity_radius)`. This is the more general version, and I guess it would make sense to use two input arguments for it.

I totally understand that the latter makes more sense from a pure programming standpoint; from a usability standpoint, I am a bit torn. We could think of two functions `rolling_ball` and `rolling_spheroid` where `rolling_ball` wraps `rolling_spheroid`? This way the function name would be decent documentation, too.

Although, at this point, it might be worthwhile to think about supporting nd-ellipsoids or arbitrary structural elements. Not sure if that would be too much generalization to be practical though. Thoughts?

Reply:
@FirefoxMetzger why not my original suggestion of "radius" and an optional "kernel_shape" which defaults to be 2x radius?

Reply:
Are nd-ellipsoids required/useful the generalization of this algorithm to ND? In case you aim to extend this to ND (I'm sure @jni would approve :wink:, I would!) I think we should hold off discussing the details of the API until that part is done.

Reply:
haha stop getting ahead of me @lagru ;) I'm happy to make this isotropic-only to begin with, with the possibility of extending to anisotropy in the future.

Reply:
@jni That one is also possible. My thoughts here are (1) that `kernel_shape` suggests that a kernel that isn't a circle is possible, and (2) that it is not immediately obvious that `radius` controls the intensity dimension, and `kernel_shape` controls the spacial dimension. For (1) this can be made possible, but currently isn't; it's essentially moving from a spheroid to an ellipsoid as a structuring element. for (2) `radius` isn't quite accurate for the intensity dimension either, because the object being "rolled" is not the cap of a ball, but actually the cap of an ellipsis. Hence `radius` specifies the (co-)vertex of an ellipsis rather than a radius.

@lagru Yes, they would be required. The logic is (roughly) the following: We start with rolling a ball/sphere in the image, but this is not general, because the intensity dimension may have a different scale than the spacial dimensions of the image. Hence we "squish" the intensity dimension (current implementation [firefoxmetzger/scikit-image@730d809]); now we have a ball/circle in the spacial dimensions with a little cap on top for the intensity dimension. If we spin that around the intensity dimension the result is a spheroid. To make it even more general, there is no need for both spacial dimensions to be scaled in the same way (this is essentially @jni 's `kernel_shape` suggestion), and the resulting object we roll in the image is an ellipsoid. This needs me to do some thinking, because it changes the equation to compute the `sagitta` (not a fitting name anymore; need to think of a better one).

Going to nd-images with multiple color channels is relatively easy; we just repeat what we do for the one intensity dimension n times (bigger matrix multiplication). The question isn't so much if this is possible to implement, but if there is practical utility. What does it mean for a hyperellipsoid/hyperspheroid to be rolled around in a (color) image? Is the result still the background intensity?